readme.md
# asgi-dispatcher-middleware

Middleware to Dispatch to multiple ASGI applications, extracted from hypercorn.

## Dispatch to multiple ASGI applications

It is often useful serve multiple ASGI applications at once, under differing root paths.
This middleware allows multiple applications to be served on different mounts.

The `DispatcherMiddleware` takes a dictionary of applications keyed by the root path.
The order of entry in this dictionary is important, as the root paths will be checked in
this order. Hence it is important to add `/a/b` before `/a` or the latter will match
everything first. Also note that the root path should not include the trailing slash.

An example usage is to to serve a graphql application alongside a static file serving
application. Using the graphql app is called `graphql_app` serving everything with the
root path `/graphql` and a static file app called `static_app` serving everything else
i.e. a root path of `/` the `DispatcherMiddleware` can be setup as,

```{.sourceCode .python}
from asgi_dispatcher_middleware import DispatcherMiddleware

dispatcher_app = DispatcherMiddleware({
    "/graphql": graphql_app,
    "/": static_app,
})
```

which can then be served by any asgi framework,

```{.sourceCode .shell}
$ hypercorn module:dispatcher_app
$ uvicorn module:dispatcher_app
```

## See also

- hypercorn: <https://pgjones.gitlab.io/hypercorn/how_to_guides/dispatch_apps.html>


changelog.md
## 1.0.1 (unreleased)

- Nothing changed yet.

## 1.0.0 (2021-01-23)

- First release on PyPI


tests/test_asgi_dispatcher_middleware.py
from __future__ import generator_stop

import pytest

from asgi_dispatcher_middleware import DispatcherMiddleware, asgitypes


def _http_scope(path: str = "/") -> asgitypes.HTTPScope:
    return {
        "type": "http",
        "asgi": {"spec_version": "2.1", "version": "3.0"},
        "http_version": "2",
        "method": "GET",
        "scheme": "https",
        "path": path,
        "raw_path": b"/",
        "query_string": b"a=b",
        "root_path": "",
        "headers": [
            (b"User-Agent", b"Hypercorn"),
            (b"X-Hypercorn", b"Hypercorn"),
            (b"Referer", b"hypercorn"),
        ],
        "client": ("127.0.0.1", 80),
        "server": None,
        "extensions": {},
    }


@pytest.mark.anyio
async def test_dispatcher_middleware() -> None:
    class EchoFramework:
        def __init__(self, name: str) -> None:
            self.name = name

        async def __call__(
            self,
            scope: asgitypes.Scope,
            receive: asgitypes.ASGIReceiveCallable,
            send: asgitypes.ASGISendCallable,
        ) -> None:
            assert scope["type"] == "http"
            response = f"{self.name}-{scope['path']}".encode()
            await send(
                asgitypes.HTTPResponseStartEvent(
                    type="http.response.start",
                    status=200,
                    headers=[(b"content-length", b"%d" % len(response))],
                )
            )
            await send(
                asgitypes.HTTPResponseBodyEvent(
                    type="http.response.body",
                    body=response,
                    more_body=False,
                )
            )

    app = DispatcherMiddleware(
        {"/api/x": EchoFramework("apix"), "/api": EchoFramework("api")}
    )

    sent_events = []

    async def send(message: object) -> None:
        sent_events.append(message)

    async def receive() -> asgitypes.ASGIReceiveEvent:
        assert False  # pragma: no cover

    await app(_http_scope(path="/api/x/b"), receive, send)
    await app(_http_scope(path="/api/b"), receive, send)
    await app(_http_scope(path="/"), receive, send)
    assert sent_events == [
        {
            "type": "http.response.start",
            "status": 200,
            "headers": [(b"content-length", b"7")],
        },
        {"type": "http.response.body", "body": b"apix-/b", "more_body": False},
        {
            "type": "http.response.start",
            "status": 200,
            "headers": [(b"content-length", b"6")],
        },
        {"type": "http.response.body", "body": b"api-/b", "more_body": False},
        {
            "type": "http.response.start",
            "status": 404,
            "headers": [(b"content-length", b"0")],
        },
        {"type": "http.response.body", "body": b"", "more_body": False},
    ]


class ScopeFramework:
    def __init__(self, name: str) -> None:
        self.name = name

    async def __call__(
        self,
        scope: asgitypes.Scope,
        receive: asgitypes.ASGIReceiveCallable,
        send: asgitypes.ASGISendCallable,
    ) -> None:
        await send(
            asgitypes.LifespanStartupCompleteEvent(type="lifespan.startup.complete")
        )


@pytest.mark.anyio
async def test_dispatcher_lifespan() -> None:
    app = DispatcherMiddleware(
        {"/apix": ScopeFramework("apix"), "/api": ScopeFramework("api")}
    )

    sent_events = []

    async def send(message: object) -> None:
        sent_events.append(message)

    async def receive() -> asgitypes.LifespanShutdownEvent:
        return {"type": "lifespan.shutdown"}

    await app(
        {"type": "lifespan", "asgi": {"spec_version": "3.0", "version": "3.0"}},
        receive,
        send,
    )
    assert sent_events == [{"type": "lifespan.startup.complete"}]


asgi_dispatcher_middleware/asgitypes.py
# copied from asgiref.typing
#
# Copyright (c) Django Software Foundation and individual contributors.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:
#
#     1. Redistributions of source code must retain the above copyright notice,
#        this list of conditions and the following disclaimer.
#
#     2. Redistributions in binary form must reproduce the above copyright
#        notice, this list of conditions and the following disclaimer in the
#        documentation and/or other materials provided with the distribution.
#
#     3. Neither the name of Django nor the names of its contributors may be used
#        to endorse or promote products derived from this software without
#        specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from __future__ import generator_stop

import sys
from typing import Awaitable, Callable, Dict, Iterable, Optional, Tuple, Type, Union

if sys.version_info >= (3, 8):
    from typing import Literal, Protocol, TypedDict
else:
    from typing_extensions import Literal, Protocol, TypedDict


class ASGIVersions(TypedDict):
    spec_version: str
    version: Union[Literal["2.0"], Literal["3.0"]]


class HTTPScope(TypedDict):
    type: Literal["http"]
    asgi: ASGIVersions
    http_version: str
    method: str
    scheme: str
    path: str
    raw_path: bytes
    query_string: bytes
    root_path: str
    headers: Iterable[Tuple[bytes, bytes]]
    client: Optional[Tuple[str, int]]
    server: Optional[Tuple[str, Optional[int]]]
    extensions: Dict[str, Dict[object, object]]


class WebsocketScope(TypedDict):
    type: Literal["websocket"]
    asgi: ASGIVersions
    http_version: str
    scheme: str
    path: str
    raw_path: bytes
    query_string: bytes
    root_path: str
    headers: Iterable[Tuple[bytes, bytes]]
    client: Optional[Tuple[str, int]]
    server: Optional[Tuple[str, Optional[int]]]
    subprotocols: Iterable[str]
    extensions: Dict[str, Dict[object, object]]


class LifespanScope(TypedDict):
    type: Literal["lifespan"]
    asgi: ASGIVersions


WWWScope = Union[HTTPScope, WebsocketScope]
Scope = Union[HTTPScope, WebsocketScope, LifespanScope]


class HTTPRequestEvent(TypedDict):
    type: Literal["http.request"]
    body: bytes
    more_body: bool


class HTTPResponseStartEvent(TypedDict):
    type: Literal["http.response.start"]
    status: int
    headers: Iterable[Tuple[bytes, bytes]]


class HTTPResponseBodyEvent(TypedDict):
    type: Literal["http.response.body"]
    body: bytes
    more_body: bool


class HTTPServerPushEvent(TypedDict):
    type: Literal["http.response.push"]
    path: str
    headers: Iterable[Tuple[bytes, bytes]]


class HTTPDisconnectEvent(TypedDict):
    type: Literal["http.disconnect"]


class WebsocketConnectEvent(TypedDict):
    type: Literal["websocket.connect"]


class WebsocketAcceptEvent(TypedDict):
    type: Literal["websocket.accept"]
    subprotocol: Optional[str]
    headers: Iterable[Tuple[bytes, bytes]]


class WebsocketReceiveEvent(TypedDict):
    type: Literal["websocket.receive"]
    bytes: Optional[bytes]
    text: Optional[str]


class WebsocketSendEvent(TypedDict):
    type: Literal["websocket.send"]
    bytes: Optional[bytes]
    text: Optional[str]


class WebsocketResponseStartEvent(TypedDict):
    type: Literal["websocket.http.response.start"]
    status: int
    headers: Iterable[Tuple[bytes, bytes]]


class WebsocketResponseBodyEvent(TypedDict):
    type: Literal["websocket.http.response.body"]
    body: bytes
    more_body: bool


class WebsocketDisconnectEvent(TypedDict):
    type: Literal["websocket.disconnect"]
    code: int


class WebsocketCloseEvent(TypedDict):
    type: Literal["websocket.close"]
    code: int


class LifespanStartupEvent(TypedDict):
    type: Literal["lifespan.startup"]


class LifespanShutdownEvent(TypedDict):
    type: Literal["lifespan.shutdown"]


class LifespanStartupCompleteEvent(TypedDict):
    type: Literal["lifespan.startup.complete"]


class LifespanStartupFailedEvent(TypedDict):
    type: Literal["lifespan.startup.failed"]
    message: str


class LifespanShutdownCompleteEvent(TypedDict):
    type: Literal["lifespan.shutdown.complete"]


class LifespanShutdownFailedEvent(TypedDict):
    type: Literal["lifespan.shutdown.failed"]
    message: str


ASGIReceiveEvent = Union[
    HTTPRequestEvent,
    HTTPDisconnectEvent,
    WebsocketConnectEvent,
    WebsocketReceiveEvent,
    WebsocketDisconnectEvent,
    LifespanStartupEvent,
    LifespanShutdownEvent,
]


ASGISendEvent = Union[
    HTTPResponseStartEvent,
    HTTPResponseBodyEvent,
    HTTPServerPushEvent,
    HTTPDisconnectEvent,
    WebsocketAcceptEvent,
    WebsocketSendEvent,
    WebsocketResponseStartEvent,
    WebsocketResponseBodyEvent,
    WebsocketCloseEvent,
    LifespanStartupCompleteEvent,
    LifespanStartupFailedEvent,
    LifespanShutdownCompleteEvent,
    LifespanShutdownFailedEvent,
]


ASGIReceiveCallable = Callable[[], Awaitable[ASGIReceiveEvent]]
ASGISendCallable = Callable[[ASGISendEvent], Awaitable[None]]


class ASGI2Protocol(Protocol):
    def __init__(self, scope: Scope) -> None:
        ...

    async def __call__(
        self, receive: ASGIReceiveCallable, send: ASGISendCallable
    ) -> None:
        ...


ASGI2Application = Type[ASGI2Protocol]
ASGI3Application = Callable[
    [
        Scope,
        ASGIReceiveCallable,
        ASGISendCallable,
    ],
    Awaitable[None],
]
ASGIApplication = Union[ASGI2Application, ASGI3Application]


asgi_dispatcher_middleware/_core.py
from __future__ import generator_stop

import functools
import inspect
import typing

import anyio

from . import asgitypes

MAX_QUEUE_SIZE = 10

T = typing.TypeVar("T")
U = typing.TypeVar("U")


class Left(typing.Generic[T, U]):
    def __init__(self, v: T):
        self.left = v


class Right(typing.Generic[T, U]):
    def __init__(self, v: U):
        self.right = v


Either = typing.Union[Left[T, U], Right[T, U]]


async def _invoke_asgi(
    app: asgitypes.ASGIApplication,
    scope: asgitypes.Scope,
    receive: asgitypes.ASGIReceiveCallable,
    send: asgitypes.ASGISendCallable,
) -> None:
    disambiguated = _disambiguate_app(app)
    if isinstance(disambiguated, Right):
        scope["asgi"]["version"] = "2.0"
        asgi_instance = disambiguated.right(scope)
        await asgi_instance(receive, send)
        return
    scope["asgi"]["version"] = "3.0"
    await disambiguated.left(scope, receive, send)


def _disambiguate_app(
    app: asgitypes.ASGIApplication,
) -> Either[asgitypes.ASGI3Application, asgitypes.ASGI2Application]:
    if inspect.isclass(app):
        return Right(typing.cast(asgitypes.ASGI2Application, app))

    if inspect.iscoroutinefunction(
        getattr(app, "__call__", None)
    ) or inspect.iscoroutinefunction(app):
        return Left(typing.cast(asgitypes.ASGI3Application, app))

    return Right(typing.cast(asgitypes.ASGI2Application, app))


class DispatcherMiddleware:
    def __init__(self, mounts: typing.Dict[str, asgitypes.ASGIApplication]) -> None:
        self.mounts = mounts

    async def __call__(
        self,
        scope: asgitypes.Scope,
        receive: asgitypes.ASGIReceiveCallable,
        send: asgitypes.ASGISendCallable,
    ) -> None:
        if scope["type"] == "lifespan":
            await self._handle_lifespan(scope, receive, send)
        else:
            for path, app in self.mounts.items():
                if scope["path"].startswith(path):
                    scope["path"] = scope["path"][len(path) :] or "/"  # type: ignore
                    return await _invoke_asgi(app, scope, receive, send)
            await send(
                asgitypes.HTTPResponseStartEvent(
                    type="http.response.start",
                    status=404,
                    headers=[(b"content-length", b"0")],
                )
            )
            await send(
                asgitypes.HTTPResponseBodyEvent(
                    type="http.response.body", body=b"", more_body=False
                )
            )

    async def _handle_lifespan(
        self,
        scope: asgitypes.Scope,
        receive: asgitypes.ASGIReceiveCallable,
        send: asgitypes.ASGISendCallable,
    ) -> None:
        self.app_queues = {
            path: anyio.create_memory_object_stream(MAX_QUEUE_SIZE)
            for path in self.mounts
        }
        self.startup_complete = {path: False for path in self.mounts}
        self.shutdown_complete = {path: False for path in self.mounts}

        async with anyio.create_task_group() as tg:
            for path, app in self.mounts.items():
                await tg.spawn(
                    _invoke_asgi,
                    app,
                    scope,
                    self.app_queues[path][1].receive,
                    functools.partial(self.send, path, send),
                )

            while True:
                message = await receive()
                for channels in self.app_queues.values():
                    await channels[0].send(message)
                if message["type"] == "lifespan.shutdown":
                    break

    async def send(
        self,
        path: str,
        send: asgitypes.ASGISendCallable,
        message: asgitypes.ASGISendEvent,
    ) -> None:
        if message["type"] == "lifespan.startup.complete":
            self.startup_complete[path] = True
            if all(self.startup_complete.values()):
                await send(
                    asgitypes.LifespanStartupCompleteEvent(
                        type="lifespan.startup.complete"
                    )
                )
        elif message["type"] == "lifespan.shutdown.complete":
            self.shutdown_complete[path] = True
            if all(self.shutdown_complete.values()):
                await send(
                    asgitypes.LifespanShutdownCompleteEvent(
                        type="lifespan.shutdown.complete"
                    )
                )
